diff --git a/HeishaMon/HeishaMon.ino b/HeishaMon/HeishaMon.ino
index ad16a39..72a5e8a 100644
--- a/HeishaMon/HeishaMon.ino
+++ b/HeishaMon/HeishaMon.ino
@@ -1,7 +1,15 @@
 #define LWIP_INTERNAL
 
+// --- build switch: classic ESP32 (bez ETH/W5500) ---
+#define HEISHAMON_NO_ETH
+
+#ifndef WL_STOPPED
+  #define WL_STOPPED WL_DISCONNECTED
+#endif
+
 #if defined(ESP8266)
 #include <ESP8266WiFi.h>
+
 #include <ESP8266mDNS.h>
   #define heatpumpSerial Serial
   #define loggingSerial Serial1
@@ -19,13 +27,19 @@
   #define PROXYTX 8
   #define ENABLEPIN 5
   #define ENABLEOTPIN 4
-  #define LEDPIN 42
+  //#define LEDPIN 42
+  #define LEDPIN 2
   #define BOOTPIN 0
 #include <WiFi.h>
 #include <ESPmDNS.h>
 #include <Adafruit_NeoPixel.h>
 #endif
 
+#if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
+static bool g_ethStarted = false;
+#endif
+
+
 
 #include <DNSServer.h>
 #include <WiFiUdp.h>
@@ -61,10 +75,14 @@ settingsStruct heishamonSettings;
 
 uint32_t neoPixelState = 0; //running neoPixelState
 bool inSetup; //bool to check if still booting
-bool sending = false; // mutex for sending data
+bool sending = false;
+
+// mutex for sending data
 bool mqttcallbackinprogress = false; // mutex for processing mqtt callback
 
-bool extraDataBlockAvailable = false; // this will be set to true if, during boot, heishamon detects this heatpump has extra data block (like K and L series do)
+bool extraDataBlockAvailable = false;
+
+// this will be set to true if, during boot, heishamon detects this heatpump has extra data block (like K and L series do)
 
 #define MQTTRECONNECTTIMER 30000 //it takes 30 secs for each mqtt server reconnect attempt
 unsigned long lastMqttReconnectAttempt = 0;
@@ -73,18 +91,24 @@ unsigned long bootButtonNotPressed = 0;
 
 #define WIFIRETRYTIMER 15000 // switch between hotspot and configured SSID each 10 secs if SSID is lost
 unsigned long lastWifiRetryTimer = 0;
+
 bool doInitialWifiScan = true; //we want an initial wifi scan to fill in the dropbox on the wifi settings page
 
 unsigned long lastRunTime = 0;
+
 unsigned long lastOptionalPCBRunTime = 0;
 unsigned long lastOptionalPCBSave = 0;
 
-unsigned long sendCommandReadTime = 0; //set to millis value during send, allow to wait millis for answer
+unsigned long sendCommandReadTime = 0;
+
+//set to millis value during send, allow to wait millis for answer
 unsigned long goodreads = 0;
+
 unsigned long totalreads = 0;
 unsigned long badcrcread = 0;
 unsigned long badheaderread = 0;
 unsigned long tooshortread = 0;
+
 unsigned long toolongread = 0;
 unsigned long timeoutread = 0;
 float readpercentage = 0;
@@ -93,12 +117,14 @@ static int uploadpercentage = 0;
 // instead of passing array pointers between functions we just define this in the global scope
 #define MAXDATASIZE 255
 char data[MAXDATASIZE] = { '\0' };
+
 byte data_length = 0;
 
 #ifdef ESP32
 //for received proxied data
 char proxydata[MAXDATASIZE] = { '\0' };
 byte proxydata_length = 0;
+
 //for the neopixel
 Adafruit_NeoPixel pixels(1, LEDPIN);
 #endif
@@ -106,6 +132,7 @@ Adafruit_NeoPixel pixels(1, LEDPIN);
 // store actual data
 char actData[DATASIZE] = { '\0' };
 char actDataExtra[DATASIZE] = { '\0' };
+
 char actOptData[OPTDATASIZE]  = { '\0' };
 
 // log message to sprintf to
@@ -122,47 +149,45 @@ static int mqttReconnects = 0;
 // buffer for commands to send
 struct cmdbuffer_t {
   uint8_t length;
-  byte data[128];
+
+byte data[128];
 } cmdbuffer[MAXCOMMANDSINBUFFER];
 
 static uint8_t cmdstart = 0;
 static uint8_t cmdend = 0;
 static uint8_t cmdnrel = 0;
 
-
-
 // mqtt
 WiFiClient mqtt_wifi_client;
 PubSubClient mqtt_client;
 
-
-
 bool firstConnectSinceBoot = true; //if this is true there is no first connection made yet
 
 struct timerqueue_t **timerqueue = NULL;
+
 int timerqueue_size = 0;
 
-#ifdef ESP32
-#define ETH_TYPE        ETH_PHY_W5500
-#define ETH_ADDR         1
-#define ETH_CS          10
-#define ETH_IRQ          15
-#define ETH_RST          14
-
-// SPI pins
-#define ETH_SPI_SCK     12
-#define ETH_SPI_MISO    13
-#define ETH_SPI_MOSI    11
-
-void setupETH() {
-  SPI.begin(ETH_SPI_SCK, ETH_SPI_MISO, ETH_SPI_MOSI);
-  if (ETH.begin(ETH_TYPE, ETH_ADDR, ETH_CS, ETH_IRQ, ETH_RST, SPI)) {
-    //sethostname on ESP32 after eth.begin (!! for wifi is most be before...!!)
-    ETH.setHostname(heishamonSettings.wifi_hostname);
-  } else {
-    loggingSerial.println("Could not start ethernet. No ethernet module installed?");
+#if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
+
+  #define ETH_TYPE ETH_PHY_W5500
+  #define ETH_ADDR 1
+  #define ETH_CS 10
+  #define ETH_IRQ 15
+  #define ETH_RST 14
+
+  #define ETH_SPI_SCK 12
+  #define ETH_SPI_MISO 13
+  #define ETH_SPI_MOSI 11
+
+  void setupETH() {
+    SPI.begin(ETH_SPI_SCK, ETH_SPI_MISO, ETH_SPI_MOSI);
+    if (ETH.begin(ETH_TYPE, ETH_ADDR, ETH_CS, ETH_IRQ, ETH_RST, SPI)) {
+      ETH.setHostname(heishamonSettings.wifi_hostname);
+    } else {
+      loggingSerial.println("Could not start ethernet. No ethernet module installed?");
+    }
   }
-}
+
 #endif
 
 
@@ -172,6 +197,7 @@ void setupETH() {
 */
 void check_wifi() {
   int wifistatus = WiFi.status();
+
   if ((wifistatus != WL_CONNECTED) && (WiFi.localIP())) {
     // special case where it seems that we are not connect but we do have working IP (causing the -1% wifi signal), do a reset.
 #ifdef ESP8266
@@ -187,8 +213,11 @@ void check_wifi() {
     */
     if (heishamonSettings.hotspot) {
       dnsServer.processNextRequest();
+
 #ifdef ESP32
-      neoPixelState = pixels.Color(16, 16, 0);  //set neopixel to yellow to indicate lost wifi
+      neoPixelState = pixels.Color(16, 16, 0);
+
+//set neopixel to yellow to indicate lost wifi
 #endif
     }
 
@@ -198,47 +227,57 @@ void check_wifi() {
 #ifdef ESP8266
     if ((heishamonSettings.wifi_ssid[0] != '\0') && (wifistatus != WL_DISCONNECTED) && (WiFi.scanComplete() != -1) && (WiFi.softAPgetStationNum() > 0)) {
       log_message(_F("WiFi lost, but softAP station connecting, so stop trying to connect to configured ssid..."));
-      WiFi.disconnect(true);
+
+WiFi.disconnect(true);
     }
 #else
     if ((heishamonSettings.wifi_ssid[0] != '\0') && (wifistatus != WL_STOPPED) && (WiFi.scanComplete() != -1) && (WiFi.softAPgetStationNum() > 0)) {
       log_message(_F("WiFi lost, but softAP station connecting, so stop trying to connect to configured ssid..."));
-      WiFi.mode(WIFI_AP);
+
+WiFi.mode(WIFI_AP);
     }
 #endif
 
-    /*  only start this routine if timeout on
+    /* only start this routine if timeout on
         reconnecting to AP and SSID is set
     */
     if ((heishamonSettings.wifi_ssid[0] != '\0') && ((unsigned long)(millis() - lastWifiRetryTimer) > WIFIRETRYTIMER)) {
       lastWifiRetryTimer = millis();
+
 #ifdef ESP8266
       if ((WiFi.softAPSSID() == "") && (heishamonSettings.hotspot)) {
         log_message(_F("WiFi lost, starting setup hotspot..."));
-        WiFi.softAPConfig(apIP, apIP, IPAddress(255, 255, 255, 0));
+
+WiFi.softAPConfig(apIP, apIP, IPAddress(255, 255, 255, 0));
         WiFi.softAP(_F("HeishaMon-Setup"));
       }
       if ((wifistatus == WL_DISCONNECTED) && (WiFi.softAPgetStationNum() == 0)) {
         log_message(_F("Retrying configured WiFi, ..."));
+
 #else
       if (((WiFi.getMode() & WIFI_MODE_AP) == 0) && (heishamonSettings.hotspot)) {
         log_message(_F("WiFi lost, starting setup hotspot..."));
-        WiFi.softAPConfig(apIP, apIP, IPAddress(255, 255, 255, 0));
+
+WiFi.softAPConfig(apIP, apIP, IPAddress(255, 255, 255, 0));
         WiFi.softAP(_F("HeishaMon-Setup"));
       }
       if ((wifistatus == WL_STOPPED  ) && (WiFi.softAPgetStationNum() == 0)) { //make sure we start STA again if it was stopped
         log_message(_F("Retrying configured WiFi, ..."));
-        WiFi.setScanMethod(WIFI_ALL_CHANNEL_SCAN); //select best AP with same SSID
+
+WiFi.setScanMethod(WIFI_ALL_CHANNEL_SCAN); //select best AP with same SSID
 #endif
         if (heishamonSettings.wifi_password[0] == '\0') {
           WiFi.begin(heishamonSettings.wifi_ssid);
-        } else {
+
+} else {
           WiFi.begin(heishamonSettings.wifi_ssid, heishamonSettings.wifi_password);
-        }
+
+}
 #ifdef ESP8266        
       } else {
         log_message(_F("Reconnecting to WiFi failed. Waiting a few seconds before trying again."));
-        WiFi.disconnect(true);
+
+WiFi.disconnect(true);
 #endif        
       }
     }
@@ -247,14 +286,15 @@ void check_wifi() {
   if (WiFi.localIP()) {  //WiFi connected
     if (WiFi.softAPSSID() != "") {
       log_message(_F("WiFi (re)connected, shutting down hotspot..."));
-      WiFi.softAPdisconnect(true);
+
+WiFi.softAPdisconnect(true);
       MDNS.notifyAPChange();
     }
 #else
-  if (WiFi.localIP() || ETH.hasIP()) {      //WiFi or ETH connected and IP>0  check if active AP and disable if yes
-    neoPixelState = pixels.Color(0, 0, 0);  //neopixel should be black again to indicate normale working      
-    if (((WiFi.getMode() & WIFI_MODE_AP) != 0) && ((heishamonSettings.wifi_ssid[0] != '\0') || (!heishamonSettings.hotspot)) ) { //shutdown hotspot if it is running with configured SSID or if hotspot config is disabled
-      log_message(_F("WiFi or ETH (re)connected, shutting down hotspot..."));
+  if (WiFi.localIP()) {      //WiFi connected and IP>0
+    neoPixelState = pixels.Color(0, 0, 0);
+    if (((WiFi.getMode() & WIFI_MODE_AP) != 0) && ((heishamonSettings.wifi_ssid[0] != '\0') || (!heishamonSettings.hotspot)) ) {
+      log_message(_F("WiFi (re)connected, shutting down hotspot..."));
       WiFi.softAP("");
       WiFi.softAPdisconnect(true);
     }
@@ -262,27 +302,34 @@ void check_wifi() {
 
     if (firstConnectSinceBoot) {  // this should start only when softap is down or else it will not work properly so run after the routine to disable softap
       firstConnectSinceBoot = false;
-      lastMqttReconnectAttempt = 0;  //initiate mqtt connection asap
+
+lastMqttReconnectAttempt = 0;  //initiate mqtt connection asap
       setupOTA();
       MDNS.begin(heishamonSettings.wifi_hostname);
       MDNS.addService("http", "tcp", 80);
+
 #ifdef ESP8266
       experimental::ESP8266WiFiGratuitous::stationKeepAliveSetIntervalMs(5000);  //necessary for some users with bad wifi routers
 #endif
 
       if (heishamonSettings.wifi_ssid[0] == '\0') {
         log_message(_F("WiFi connected without SSID and password in settings. Must come from persistent memory. Storing in settings."));
-        WiFi.SSID().toCharArray(heishamonSettings.wifi_ssid, 40);
+
+WiFi.SSID().toCharArray(heishamonSettings.wifi_ssid, 40);
         WiFi.psk().toCharArray(heishamonSettings.wifi_password, 40);
         JsonDocument jsonDoc;
         settingsToJson(jsonDoc, &heishamonSettings);  //stores current settings in a json document
-        saveJsonToFile(jsonDoc, "config.json");     //save to config file
+        saveJsonToFile(jsonDoc, "config.json");
+
+//save to config file
       }
 
       ntpReload(&heishamonSettings);
-      logprintln_P(F("Try to syncing with ntp servers. Checking again in 5 minutes"));
+
+logprintln_P(F("Try to syncing with ntp servers. Checking again in 5 minutes"));
       timerqueue_insert(300, 0, -6);
-    }
+
+}
 
     /*
        always update if wifi is working so next time on ssid failure
@@ -293,15 +340,18 @@ void check_wifi() {
 #ifdef ESP8266
     // Allow MDNS processing
     MDNS.update();
+
 #endif
   }
   if (doInitialWifiScan && (millis() > 15000)) {  //do a wifi scan a boot after 15 seconds
     doInitialWifiScan = false;
-    log_message(_F("Starting initial wifi scan ..."));
+
+log_message(_F("Starting initial wifi scan ..."));
 #if defined(ESP8266)
     WiFi.scanNetworksAsync(getWifiScanResults);
 #elif defined(ESP32)
     WiFi.scanNetworks(true);
+
 #endif
   }
 }
@@ -310,56 +360,63 @@ void check_wifi() {
 void mqtt_reconnect()
 {
   unsigned long now = millis();
-  if ((lastMqttReconnectAttempt == 0) || ((unsigned long)(now - lastMqttReconnectAttempt) > MQTTRECONNECTTIMER)) { //only try reconnect each MQTTRECONNECTTIMER seconds or on boot when lastMqttReconnectAttempt is still 0
+
+if ((lastMqttReconnectAttempt == 0) || ((unsigned long)(now - lastMqttReconnectAttempt) > MQTTRECONNECTTIMER)) { //only try reconnect each MQTTRECONNECTTIMER seconds or on boot when lastMqttReconnectAttempt is still 0
     lastMqttReconnectAttempt = now;
-    if (mqttReconnects == 0) {
+
+if (mqttReconnects == 0) {
       log_message(_F("Connecting to mqtt server ..."));
-    } else {
+
+} else {
       log_message(_F("Reconnecting to mqtt server ..."));
     }
     char topic[256];
-    sprintf(topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_willtopic);
+
+sprintf(topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_willtopic);
     if (mqtt_client.connect(heishamonSettings.wifi_hostname, heishamonSettings.mqtt_username, heishamonSettings.mqtt_password, topic, 1, true, "Offline"))
     {
       mqttReconnects++;
-      if (heishamonSettings.opentherm) {
+
+if (heishamonSettings.opentherm) {
         sprintf(topic, "%s/%s/#", heishamonSettings.mqtt_topic_base, mqtt_topic_opentherm_read);
         mqtt_client.subscribe(topic);
-      }
+
+}
       sprintf(topic, "%s/%s/#", heishamonSettings.mqtt_topic_base, mqtt_topic_commands);
       mqtt_client.subscribe(topic);
       sprintf(topic, "%s/%s/#", heishamonSettings.mqtt_topic_base, mqtt_topic_gpio);
       mqtt_client.subscribe(topic);      
       sprintf(topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_send_raw_value_topic);
-      mqtt_client.subscribe(topic);
+
+mqtt_client.subscribe(topic);
       sprintf(topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_willtopic);
       mqtt_client.publish(topic, "Online");
       sprintf(topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_iptopic);
-#ifdef ESP8266
-      mqtt_client.publish(topic, WiFi.localIP().toString().c_str(), true);
-#else
-      if (ETH.hasIP()) {
-        mqtt_client.publish(topic, ETH.localIP().toString().c_str(), true);
-      } else {
-        mqtt_client.publish(topic, WiFi.localIP().toString().c_str(), true);
-      }
-#endif
 
-      if (heishamonSettings.use_s0) { // connect to s0 topic to retrieve older watttotal from mqtt
+mqtt_client.publish(topic, WiFi.localIP().toString().c_str(), true);
+
+if (heishamonSettings.use_s0) { // connect to s0 topic to retrieve older watttotal from mqtt
         sprintf_P(mqtt_topic, PSTR("%s/%s/WatthourTotal/1"), heishamonSettings.mqtt_topic_base, mqtt_topic_s0);
-        mqtt_client.subscribe(mqtt_topic);
+
+mqtt_client.subscribe(mqtt_topic);
         sprintf_P(mqtt_topic, PSTR("%s/%s/WatthourTotal/2"), heishamonSettings.mqtt_topic_base, mqtt_topic_s0);
         mqtt_client.subscribe(mqtt_topic);
       }
       if (mqttReconnects == 1) { //only resend all data on first connect to mqtt so a data bomb like and bad mqtt server will not cause a reconnect bomb everytime
-        if (heishamonSettings.use_1wire) resetlastalldatatime_dallas(); //resend all 1wire values to mqtt
-        resetlastalldatatime(); //resend all heatpump values to mqtt
+        if (heishamonSettings.use_1wire) resetlastalldatatime_dallas();
+
+//resend all 1wire values to mqtt
+        resetlastalldatatime();
+
+//resend all heatpump values to mqtt
       }
       //use this to receive valid heishamon raw data from other heishamon to debug this OT code
 //#define RAWDEBUG
 #ifdef RAWDEBUG
       if ( heishamonSettings.listenonly) {
-        mqtt_client.subscribe((char*)"panasonic_heat_pump/raw/data"); //subscribe to raw heatpump data over MQTT
+        mqtt_client.subscribe((char*)"panasonic_heat_pump/raw/data");
+
+//subscribe to raw heatpump data over MQTT
       }
 #endif
     }
@@ -369,11 +426,14 @@ void mqtt_reconnect()
 #ifdef ESP32
 void blinkNeoPixel(bool status) {
   if (status) {
-    pixels.setPixelColor(0, 0, 0, 16); //blue
+    pixels.setPixelColor(0, 0, 0, 16);
+
+//blue
   } else {
     pixels.setPixelColor(0, neoPixelState);
   }
-  pixels.show(); 
+  pixels.show();
+
 }
 #endif  
 
@@ -385,34 +445,42 @@ void log_message(char* string)
 #endif
   time_t rawtime;
   rawtime = time(NULL);
-  struct tm *timeinfo = localtime(&rawtime);
+
+struct tm *timeinfo = localtime(&rawtime);
   char timestring[32];
   strftime(timestring, 32, "%c", timeinfo);
-  size_t len = strlen(string) + strlen(timestring) + 32; //+32 long enough to contain millis() and the json part later for websocket mesg
+  size_t len = strlen(string) + strlen(timestring) + 32;
+
+//+32 long enough to contain millis() and the json part later for websocket mesg
   char* log_line = (char *) malloc(len);
-  snprintf(log_line, len, "%s (%lu): %s", timestring, millis(), string);
+
+snprintf(log_line, len, "%s (%lu): %s", timestring, millis(), string);
 
   if (heishamonSettings.logSerial1) {
     loggingSerial.println(log_line);
-  }
+
+}
   if (heishamonSettings.logMqtt && mqtt_client.connected())
   {
     char log_topic[256];
     sprintf(log_topic, "%s/%s", heishamonSettings.mqtt_topic_base, mqtt_logtopic);
 
-    if (!mqtt_client.publish(log_topic, log_line)) {
+if (!mqtt_client.publish(log_topic, log_line)) {
       if (heishamonSettings.logSerial1) {
         loggingSerial.print(millis());
-        loggingSerial.print(F(": "));
+
+loggingSerial.print(F(": "));
         loggingSerial.println(F("MQTT publish log message failed!"));
       }
       mqtt_client.disconnect();
-    }
+
+}
   }
   //send log message to websocket
   snprintf(log_line, len+12, "{\"logMsg\":\"%s (%lu): %s\"}", timestring, millis(), string);
   websocket_write_all(log_line, strlen(log_line));
-  free(log_line);
+
+free(log_line);
 #ifdef ESP32
   if (!inSetup) blinkNeoPixel(false);
 #endif  
@@ -422,40 +490,50 @@ void logHex(char *hex, byte hex_len) {
 #define LOGHEXBYTESPERLINE 32  // please be aware of max mqtt message size
   for (int i = 0; i < hex_len; i += LOGHEXBYTESPERLINE) {
     char buffer [(LOGHEXBYTESPERLINE * 3) + 1];
-    buffer[LOGHEXBYTESPERLINE * 3] = '\0';
+
+buffer[LOGHEXBYTESPERLINE * 3] = '\0';
     for (int j = 0; ((j < LOGHEXBYTESPERLINE) && ((i + j) < hex_len)); j++) {
       sprintf(&buffer[3 * j], "%02X ", hex[i + j]);
-    }
+
+}
     sprintf_P(log_msg, PSTR("data: %s"), buffer ); log_message(log_msg);
-  }
+
+}
 }
 
 void mqttPublish(char* topic, char* subtopic, char* value) {
   mqttPublish(topic, subtopic, value, MQTT_RETAIN_VALUES);
+
 }
 
 void mqttPublish(char* topic, char* subtopic, char* value, bool retain) {
   char mqtt_topic[256];
   sprintf_P(mqtt_topic, PSTR("%s/%s/%s"), heishamonSettings.mqtt_topic_base, topic, subtopic);
-  mqtt_client.publish(mqtt_topic, value, retain);
+
+mqtt_client.publish(mqtt_topic, value, retain);
 }
 
 
 
 byte calcChecksum(byte* command, int length) {
   byte chk = 0;
-  for ( int i = 0; i < length; i++)  {
+
+for ( int i = 0; i < length; i++)  {
     chk += command[i];
-  }
+
+}
   chk = (chk ^ 0xFF) + 01;
   return chk;
+
 }
 
 bool isValidReceiveChecksum(char* check_data, byte check_length) {
   byte chk = 0;
-  for ( int i = 0; i < check_length; i++)  {
+
+for ( int i = 0; i < check_length; i++)  {
     chk += check_data[i];
-  }
+
+}
   return (chk == 0); //all received bytes + checksum should result in 0
 }
 
@@ -463,69 +541,99 @@ bool isValidReceiveChecksum(char* check_data, byte check_length) {
 void readProxy()
 {
   int proxylen = 0;
-  while ((proxySerial.available()) && ((proxydata_length + proxylen) < MAXDATASIZE)) {
-    proxydata[proxydata_length + proxylen] = proxySerial.read(); //read available data and place it after the last received data
+
+while ((proxySerial.available()) && ((proxydata_length + proxylen) < MAXDATASIZE)) {
+    proxydata[proxydata_length + proxylen] = proxySerial.read();
+
+//read available data and place it after the last received data
     proxylen++;
-    if ((proxydata[0] != 0x71) and  (proxydata[0] != 0x31) and  (proxydata[0] != 0xF1)) { //wrong header received!
-      log_message(_F("PROXY Received bad header. Ignoring this data!"));
+
+if ((proxydata[0] != 0x71) and  (proxydata[0] != 0x31) and  (proxydata[0] != 0xF1)) { //wrong header received!
+
+log_message(_F("PROXY Received bad header. Ignoring this data!"));
       if (heishamonSettings.logHexdump) logHex(proxydata, proxylen);
       proxydata_length = 0;
-      return; //return so this while loop does not loop forever if there happens to be a continous invalid data stream
+      return;
+
+//return so this while loop does not loop forever if there happens to be a continous invalid data stream
     }
   }
-  //if ((proxylen > 0) && (proxydata_length == 0 )) proxy_totalreads++; //this is the start of a new read
+  //if ((proxylen > 0) && (proxydata_length == 0 )) proxy_totalreads++;
+
+//this is the start of a new read
   proxydata_length +=  proxylen;
-  if (proxydata_length > 1 ) { //should have received length part of header now
+
+if (proxydata_length > 1 ) { //should have received length part of header now
     if ((proxydata_length > ( proxydata[1] + 3)) || (proxydata_length >= MAXDATASIZE)) {
       sprintf_P(log_msg, PSTR("PROXY Received %i bytes proxy %i\n"), proxydata_length, proxydata[1]);
-      log_message(log_msg);
+
+log_message(log_msg);
       log_message(_F("PROXY Received more data than header suggests! Ignoring this as this is bad data."));
       proxydata_length = 0;
-      if (heishamonSettings.logHexdump) logHex(proxydata, proxydata_length);
+
+if (heishamonSettings.logHexdump) logHex(proxydata, proxydata_length);
       return;
     }
     if (proxydata_length == (proxydata[1] + 3)) { //we received all data (serial2_data[1] is header length field)
-      sprintf_P(log_msg, PSTR("PROXY Received %i bytes"), proxydata_length); log_message(log_msg);
+      sprintf_P(log_msg, PSTR("PROXY Received %i bytes"), proxydata_length);
+
+log_message(log_msg);
       if (heishamonSettings.logHexdump) logHex(proxydata, proxydata_length);
       if (! isValidReceiveChecksum(proxydata,proxydata_length) ) {
         log_message(_F("PROXY Checksum received false!"));
-        proxydata_length = 0; //for next attempt
+
+proxydata_length = 0; //for next attempt
         return;
-      }      
+
+}      
       log_message(_F("PROXY Checksum and header received ok!"));
-      if ((proxydata[0]==0x71 or proxydata[0]==0xF1) and proxydata_length == (PANASONICQUERYSIZE+1)) { //this is a query from cztaw on proxy port
+
+if ((proxydata[0]==0x71 or proxydata[0]==0xF1) and proxydata_length == (PANASONICQUERYSIZE+1)) { //this is a query from cztaw on proxy port
         if (proxydata[0]==0xf1) {  //this is a write query, just pass this message forward as new command
           log_message(_F("PROXY received write query, copy message forward to heatpump"));
-          send_command((byte*)proxydata,proxydata_length-1); //strip CRC, will be calculated again in send_command
+
+send_command((byte*)proxydata,proxydata_length-1); //strip CRC, will be calculated again in send_command
           //then just reply with the current settings, for read and write it is the same as the write is only acknowledged in the next read
           //so we just run to the next if statement
         }
         if (proxydata[3] == 0x10) {
           log_message(_F("PROXY requests basic data"));
-          if ((actData[0] == 0x71) && (actData[1] == 0xc8) && (actData[2] == 0x01)) { //don't answer if we don't have data
-            proxySerial.write(actData,DATASIZE); //should contain valid checksum also
+
+if ((actData[0] == 0x71) && (actData[1] == 0xc8) && (actData[2] == 0x01)) { //don't answer if we don't have data
+            proxySerial.write(actData,DATASIZE);
+
+//should contain valid checksum also
           }
         } else if (proxydata[3] == 0x21 ) {
           log_message(_F("PROXY requests extra data"));
-          if ((actDataExtra[0] == 0x71) && (actDataExtra[1] == 0xc8) && (actDataExtra[2] == 0x01)) { //don't answer if we don't have data
-            proxySerial.write(actDataExtra,DATASIZE); //should containt valid checksum also
+
+if ((actDataExtra[0] == 0x71) && (actDataExtra[1] == 0xc8) && (actDataExtra[2] == 0x01)) { //don't answer if we don't have data
+            proxySerial.write(actDataExtra,DATASIZE);
+
+//should containt valid checksum also
           }
         } else {
           log_message(_F("PROXY has sent unknown query! Forwarding to heatpump!"));
-          send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
+
+send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
         }
         proxydata_length = 0;
-        return;
+
+return;
       } else if (proxydata[0]==0x31) {
         log_message(_F("PROXY received startup message, forwarding to heatpump!"));
-        send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
+
+send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
         proxydata_length = 0;
-        return;
+
+return;
       } else {
         log_message(_F("PROXY received unknown message, forwarding it to heatpump anyway!"));
-        send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
+
+send_command((byte *)proxydata, proxydata_length-1); //strip CRC from end as send_command wil recalculate it
         proxydata_length = 0;
-        return;
+
+return;
       }
     }
   }
@@ -535,101 +643,136 @@ void readProxy()
 bool readSerial()
 {
   int len = 0;
-  while ((heatpumpSerial.available()) && ((data_length + len) < MAXDATASIZE)) {
-    data[data_length + len] = heatpumpSerial.read(); //read available data and place it after the last received data
+
+while ((heatpumpSerial.available()) && ((data_length + len) < MAXDATASIZE)) {
+    data[data_length + len] = heatpumpSerial.read();
+
+//read available data and place it after the last received data
     len++;
-    if ((data[0] != 0x71) && (data[0] != 0x31)) { //wrong header received!
+
+if ((data[0] != 0x71) && (data[0] != 0x31)) { //wrong header received!
       log_message(_F("Received bad header. Ignoring this data!"));
-      if (heishamonSettings.logHexdump) logHex(data, len);
+
+if (heishamonSettings.logHexdump) logHex(data, len);
       badheaderread++;
       data_length = 0;
-      return false; //return so this while loop does not loop forever if there happens to be a continous invalid data stream
+      return false;
+
+//return so this while loop does not loop forever if there happens to be a continous invalid data stream
     }
   }
 
-  if ((len > 0) && (data_length == 0 )) totalreads++; //this is the start of a new read
+  if ((len > 0) && (data_length == 0 )) totalreads++;
+
+//this is the start of a new read
   data_length += len;
 
-  if (data_length > 1) { //should have received length part of header now
+if (data_length > 1) { //should have received length part of header now
 
     if ((data_length > (data[1] + 3)) || (data_length >= MAXDATASIZE) ) {
       log_message(_F("Received more data than header suggests! Ignoring this as this is bad data."));
-      if (heishamonSettings.logHexdump) logHex(data, data_length);
+
+if (heishamonSettings.logHexdump) logHex(data, data_length);
       data_length = 0;
       toolongread++;
       return false;
-    }
+
+}
 
     if (data_length == (data[1] + 3)) { //we received all data (data[1] is header length field)
-      sprintf_P(log_msg, PSTR("Received %d bytes data"), data_length); log_message(log_msg);
+      sprintf_P(log_msg, PSTR("Received %d bytes data"), data_length);
+
+log_message(log_msg);
       sending = false; //we received an answer after our last command so from now on we can start a new send request again
       if (heishamonSettings.logHexdump) logHex(data, data_length);
-      if (! isValidReceiveChecksum(data, data_length) ) {
+
+if (! isValidReceiveChecksum(data, data_length) ) {
         log_message(_F("Checksum received false!"));
-        data_length = 0; //for next attempt
+        data_length = 0;
+
+//for next attempt
         badcrcread++;
         return false;
-      }
+
+}
       log_message(_F("Checksum and header received ok!"));
       goodreads++;
 
-      if (data_length == DATASIZE)  {  //receive a full data block
+if (data_length == DATASIZE)  {  //receive a full data block
         if  (data[3] == 0x10) { //decode the normal data block
           decode_heatpump_data(data, actData, mqtt_client, log_message, heishamonSettings.mqtt_topic_base, heishamonSettings.updateAllTime);
-          {
+
+{
             char mqtt_topic[256];
             sprintf(mqtt_topic, "%s/raw/data", heishamonSettings.mqtt_topic_base);
-            mqtt_client.publish(mqtt_topic, (const uint8_t *)actData, DATASIZE, false); //do not retain this raw data
+
+mqtt_client.publish(mqtt_topic, (const uint8_t *)actData, DATASIZE, false); //do not retain this raw data
           }
           data_length = 0;
-          return true;
+
+return true;
         } else if (data[3] == 0x21) { //decode the new model extra data block
-          extraDataBlockAvailable = true; //set the flag to true so we know we can request this data always
+          extraDataBlockAvailable = true;
+
+//set the flag to true so we know we can request this data always
           decode_heatpump_data_extra(data, actDataExtra, mqtt_client, log_message, heishamonSettings.mqtt_topic_base, heishamonSettings.updateAllTime);
-          {
+
+{
             char mqtt_topic[256];
             sprintf(mqtt_topic, "%s/raw/dataextra", heishamonSettings.mqtt_topic_base);
-            mqtt_client.publish(mqtt_topic, (const uint8_t *)actDataExtra, DATASIZE, false); //do not retain this raw data
+
+mqtt_client.publish(mqtt_topic, (const uint8_t *)actDataExtra, DATASIZE, false); //do not retain this raw data
           }
           data_length = 0;
-          return true;
+
+return true;
         } else {
 #ifdef ESP8266
           log_message(_F("Received an unknown full size datagram. Can't decode this yet."));
+
 #else 
           log_message(_F("Received a full size datagram but not for me. Forwarding to proxy port."));
-          proxySerial.write(data,data_length);
+
+proxySerial.write(data,data_length);
 #endif               
           data_length = 0;
-          return false;
+
+return false;
         }
       }
       else if (data_length == OPTDATASIZE ) { //optional pcb acknowledge answer
         log_message(_F("Received optional PCB ack answer. Decoding this in OPT topics."));
-        decode_optional_heatpump_data(data, actOptData, mqtt_client, log_message, heishamonSettings.mqtt_topic_base, heishamonSettings.updateAllTime);
+
+decode_optional_heatpump_data(data, actOptData, mqtt_client, log_message, heishamonSettings.mqtt_topic_base, heishamonSettings.updateAllTime);
         data_length = 0;
         return true;
-      }
+
+}
       else {
 #ifdef ESP8266
         log_message(_F("Received a shorter datagram. Can't decode this yet."));
+
 #else
         log_message(_F("Received a shorter datagram but not for me. Forwarding to proxy port."));
-        proxySerial.write(data,data_length);
+
+proxySerial.write(data,data_length);
 #endif           
         data_length = 0;
-        return false;
+
+return false;
       }
     }
   }
   return false;
+
 }
 
 void popCommandBuffer() {
   // to make sure we can pop a command from the buffer
   if ((!sending) && cmdnrel > 0) {
     send_command(cmdbuffer[cmdstart].data, cmdbuffer[cmdstart].length);
-    cmdstart = (cmdstart + 1) % (MAXCOMMANDSINBUFFER);
+
+cmdstart = (cmdstart + 1) % (MAXCOMMANDSINBUFFER);
     cmdnrel--;
   }
 }
@@ -637,93 +780,128 @@ void popCommandBuffer() {
 void pushCommandBuffer(byte* command, int length) {
   if (cmdnrel + 1 > MAXCOMMANDSINBUFFER) {
     log_message(_F("Too much commands already in buffer. Ignoring this commands.\n"));
-    return;
+
+return;
   }
   cmdbuffer[cmdend].length = length;
   memcpy(&cmdbuffer[cmdend].data, command, length);
   cmdend = (cmdend + 1) % (MAXCOMMANDSINBUFFER);
   cmdnrel++;
+
 }
 
 bool send_command(byte* command, int length) {
   if ( heishamonSettings.listenonly ) {
     log_message(_F("Not sending this command. Heishamon in listen only mode!"));
-    return false;
+
+return false;
   }
   if ( sending ) {
     log_message(_F("Already sending data. Buffering this send request"));
-    pushCommandBuffer(command, length);
+
+pushCommandBuffer(command, length);
     return false;
   }
   sending = true; //simple semaphore to only allow one send command at a time, semaphore ends when answered data is received
 
   byte chk = calcChecksum(command, length);
-  int bytesSent = heatpumpSerial.write(command, length); //first send command
-  bytesSent += heatpumpSerial.write(chk); //then calculcated checksum byte afterwards
+
+int bytesSent = heatpumpSerial.write(command, length); //first send command
+  bytesSent += heatpumpSerial.write(chk);
+
+//then calculcated checksum byte afterwards
   sprintf_P(log_msg, PSTR("sent bytes: %d including checksum value: %d "), bytesSent, int(chk));
   log_message(log_msg);
 
-  if (heishamonSettings.logHexdump) logHex((char*)command, length);
+if (heishamonSettings.logHexdump) logHex((char*)command, length);
   sendCommandReadTime = millis(); //set sendCommandReadTime when to timeout the answer of this command
   return true;
+
 }
 
 // Callback function that is called when a message has been pushed to one of your topics.
+
 void mqtt_callback(char* topic, byte* payload, unsigned int length) {
   if (mqttcallbackinprogress) {
     log_message(_F("Already processing another mqtt callback. Ignoring this one"));
-  }
+
+}
   else {
-    mqttcallbackinprogress = true; //simple semaphore to make sure we don't have two callbacks at the same time
+    mqttcallbackinprogress = true;
+
+//simple semaphore to make sure we don't have two callbacks at the same time
     char msg[length + 1];
-    for (unsigned int i = 0; i < length; i++) {
+
+for (unsigned int i = 0; i < length; i++) {
       msg[i] = (char)payload[i];
-    }
+
+}
     msg[length] = '\0';
-    char* topic_command = topic + strlen(heishamonSettings.mqtt_topic_base) + 1; //strip base plus seperator from topic
+    char* topic_command = topic + strlen(heishamonSettings.mqtt_topic_base) + 1;
+
+//strip base plus seperator from topic
     if (strcmp(topic_command, mqtt_send_raw_value_topic) == 0)
     { // send a raw hex string
       byte *rawcommand;
-      rawcommand = (byte *) malloc(length);
+
+rawcommand = (byte *) malloc(length);
       memcpy(rawcommand, msg, length);
 
       sprintf_P(log_msg, PSTR("sending raw value"));
       log_message(log_msg);
       send_command(rawcommand, length);
       free(rawcommand);
-    } else if (strncmp(topic_command, mqtt_topic_s0, strlen(mqtt_topic_s0)) == 0)  // this is a s0 topic, check for watthour topic and restore it
+
+} else if (strncmp(topic_command, mqtt_topic_s0, strlen(mqtt_topic_s0)) == 0)  // this is a s0 topic, check for watthour topic and restore it
     {
-      char* topic_s0_watthour_port = topic_command + strlen(mqtt_topic_s0) + 15; //strip the first 17 "s0/WatthourTotal/" from the topic to get the s0 port
+      char* topic_s0_watthour_port = topic_command + strlen(mqtt_topic_s0) + 15;
+
+//strip the first 17 "s0/WatthourTotal/" from the topic to get the s0 port
       int s0Port = String(topic_s0_watthour_port).toInt();
-      float watthour = String(msg).toFloat();
+
+float watthour = String(msg).toFloat();
       restore_s0_Watthour(s0Port, watthour);
       //unsubscribe after restoring the watthour values
       char mqtt_topic[256];
-      sprintf(mqtt_topic, "%s", topic);
+
+sprintf(mqtt_topic, "%s", topic);
       if (mqtt_client.unsubscribe(mqtt_topic)) {
         log_message(_F("Unsubscribed from S0 watthour restore topic"));
-      }
+
+}
     } else if (strncmp(topic_command, mqtt_topic_commands, strlen(mqtt_topic_commands)) == 0)  // check for commands to heishamon
     {
-      char* topic_sendcommand = topic_command + strlen(mqtt_topic_commands) + 1; //strip the first 9 "commands/" from the topic to get what we need
+      char* topic_sendcommand = topic_command + strlen(mqtt_topic_commands) + 1;
+
+//strip the first 9 "commands/" from the topic to get what we need
       send_heatpump_command(topic_sendcommand, msg, send_command, log_message, heishamonSettings.optionalPCB);
-    //use this to receive valid heishamon raw data from other heishamon to debug this OT code
+
+//use this to receive valid heishamon raw data from other heishamon to debug this OT code
 #ifdef RAWDEBUG
     } else if (strcmp((char*)"panasonic_heat_pump/raw/data", topic) == 0) {  // check for raw heatpump input
       sprintf_P(log_msg, PSTR("Received raw heatpump data from MQTT"));
-      log_message(log_msg);
+
+log_message(log_msg);
       decode_heatpump_data(msg, actData, mqtt_client, log_message, heishamonSettings.mqtt_topic_base, heishamonSettings.updateAllTime);
       memcpy(actData, msg, DATASIZE);
+
 #endif
     } else if (strncmp(topic_command, mqtt_topic_opentherm_read, strlen(mqtt_topic_opentherm_read)) == 0)  {
-      char* topic_otcommand = topic_command + strlen(mqtt_topic_opentherm_read) + 1; //strip the opentherm subtopic from the topic
+      char* topic_otcommand = topic_command + strlen(mqtt_topic_opentherm_read) + 1;
+
+//strip the opentherm subtopic from the topic
       mqttOTCallback(topic_otcommand, msg);
-    } else if (strncmp(topic_command, mqtt_topic_gpio, strlen(mqtt_topic_gpio)) == 0)  {
-      char* topic_gpiocommand = topic_command + strlen(mqtt_topic_gpio) + 1; //strip the gpio subtopic from the topic
+
+} else if (strncmp(topic_command, mqtt_topic_gpio, strlen(mqtt_topic_gpio)) == 0)  {
+      char* topic_gpiocommand = topic_command + strlen(mqtt_topic_gpio) + 1;
+
+//strip the gpio subtopic from the topic
       mqttGPIOCallback(topic_gpiocommand, msg);
-    }    
+
+}    
     mqttcallbackinprogress = false;
-  }
+
+}
 }
 
 void setupOTA() {
@@ -733,20 +911,22 @@ void setupOTA() {
   // Hostname defaults to esp8266-[ChipID]
   ArduinoOTA.setHostname(heishamonSettings.wifi_hostname);
 
-  // Set authentication
+// Set authentication
   ArduinoOTA.setPassword(heishamonSettings.ota_password);
 
   ArduinoOTA.onStart([]() {
   });
   ArduinoOTA.onEnd([]() {
   });
-  ArduinoOTA.onProgress([](unsigned int progress, unsigned int total) {
+
+ArduinoOTA.onProgress([](unsigned int progress, unsigned int total) {
 
   });
   ArduinoOTA.onError([](ota_error_t error) {
 
   });
   ArduinoOTA.begin();
+
 }
 
 
@@ -1241,7 +1421,7 @@ void switchSerial() {
   heatpumpSerial.flush();
   proxySerial.flush();
   proxySerial.end();
-  proxySerial.begin(9600, SERIAL_8E1,PROXYRX,PROXYTX);
+  //proxySerial.begin(9600, SERIAL_8E1,PROXYRX,PROXYTX);
   proxySerial.flush();  
 #endif
 
@@ -1432,9 +1612,6 @@ void setup() {
 
   pinMode(BOOTPIN,INPUT_PULLUP); //enable the boot switch to be used as an input after booting
 
-  loggingSerial.println(F("Send current wifi info to serial..."));
-  WiFi.printDiag(loggingSerial);
-
   loggingSerial.println(F("Loading config from flash..."));
   loadSettings(&heishamonSettings);
 
@@ -1442,7 +1619,10 @@ void setup() {
   setupWifi(&heishamonSettings);
   lastWifiRetryTimer = millis();
 
-#if defined(ESP32)
+  loggingSerial.println(F("Send current wifi info to serial..."));
+  WiFi.printDiag(loggingSerial);
+
+#if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
   loggingSerial.println(F("Setup ethernet module..."));
   setupETH();
 #endif
@@ -1631,7 +1811,11 @@ void loop() {
   #ifdef ESP8266
     if ( WiFi.isConnected() && (!mqtt_client.connected()) )
   #else
-    if ( (WiFi.isConnected() || ETH.connected()) && (!mqtt_client.connected()) )
+    #if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
+      if ( (WiFi.isConnected() || ETH.connected()) && (!mqtt_client.connected()) )
+    #else
+      if ( WiFi.isConnected() && (!mqtt_client.connected()) )
+    #endif
   #endif
     {
       if (mqttReconnects > 0 ) log_message(_F("Lost MQTT connection!"));
@@ -1667,9 +1851,9 @@ void loop() {
     message += getWifiQuality();
     message += F("% (RSSI: ");
     message += WiFi.RSSI();
-#ifdef ESP32
+#if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
     message += F(") ## Ethernet: ");
-    if (ETH.phyAddr() != 0) {        
+    if (ETH.phyAddr() != 0) {
       if (ETH.connected()) {
         if (ETH.hasIP()) {
           message += F("connected (");
@@ -1678,8 +1862,7 @@ void loop() {
         } else {
           message += F("connected (no IP)");
         }
-      } 
-      else {
+      } else {
         message += F("not connected");
       }
     } else {
@@ -1745,32 +1928,19 @@ void loop() {
     mqtt_client.publish(mqtt_topic, stats.c_str(), MQTT_RETAIN_VALUES);
 
     //websocket stats
-#ifdef ESP32
-    String ethernetStat;
-    if (ETH.phyAddr() != 0) {        
-      if (ETH.connected()) {
-        if (ETH.hasIP()) {
-          ethernetStat = F("connected - IP: ");
-          ethernetStat += ETH.localIP().toString();
-          ethernetStat += F(")");
-        } else {
-          ethernetStat = F("connected - no IP");
-        }
-      } 
-      else {
-        ethernetStat = F("not connected");
-      }
-    } else {
-      ethernetStat = F("not installed");
-    }
-    char *getuptime = getUptime();
-    sprintf_P(log_msg, PSTR("{\"data\": {\"stats\": {\"wifi\": %d, \"ethernet\": \"%s\", \"memory\": %d, \"correct\": %.0f,\"mqtt\": %d,\"uptime\": \"%s\"}}}"), getWifiQuality(), ethernetStat.c_str(), getFreeMemory(), readpercentage, mqttReconnects, getuptime);
-    free(getuptime);    
+#if defined(ESP32) && !defined(HEISHAMON_NO_ETH)
+String ethernetStat;
+if (!g_ethStarted) {
+  ethernetStat = F("not installed");
+} else {
+  IPAddress ip = ETH.localIP();
+  if (ip == IPAddress(0,0,0,0)) ethernetStat = F("connected - no IP");
+  else ethernetStat = String(F("connected - IP ")) + ip.toString();
+}
 #else
-    char *getuptime = getUptime();
-    sprintf_P(log_msg, PSTR("{\"data\": {\"stats\": {\"wifi\": %d, \"memory\": %d, \"correct\": %.0f,\"mqtt\": %d,\"uptime\": \"%s\"}}}"), getWifiQuality(), getFreeMemory(), readpercentage, mqttReconnects, getuptime);    
-    free(getuptime);    
+String ethernetStat = F("disabled");
 #endif
+
     
     websocket_write_all(log_msg, strlen(log_msg));        
 
